format/ComputeIR.fbs

```diff
@@ -0,0 +1,348 @@
+}
+
+/// The contents of Relation.options will be AggregateOptions
+/// if Relation.operation = CanonicalOperation::Aggregate
+table AggregateOptions {
+  /// Expressions which will be evaluated to produce to
+  /// the rows of the aggregate relation's output.
+  aggregations: [Expression] (required);
+  /// Keys by which `aggregations` will be grouped.
+  keys: [Expression] (required);
+}
+
+enum CanonicalJoinKindId : uint32 {
+  Inner,
+  LeftOuter,
+  RightOuter,
+  FullOuter,
+  Cross,
+}
+
+table CanonicalJoinKind {
+  id: CanonicalJoinKindId;
+}
+
+table NonCanonicalJoinKind {
+  name_space: string (required);
+  name: string (required);
+}
+
+union JoinKind {
+  CanonicalJoinKind, NonCanonicalJoinKind
+}
+
+/// The contents of Relation.options will be JoinOptions
+/// if Relation.operation = CanonicalOperation::Join
+table JoinOptions {
+  /// The expression which will be evaluated against rows from each
+  /// input to determine whether they should be included in the
+  /// join relation's output.
+  on_expression: Expression (required);
+  /// The kind of join to use.
+  join_kind: JoinKind (required);
+}
+
+/// Whether lesser values should precede greater or vice versa,
+/// also whether nulls should preced or follow values.
+enum Ordering : uint8 {
+  ASCENDING_THEN_NULLS,
+  DESCENDING_THEN_NULLS,
+  NULLS_THEN_ASCENDING,
+  NULLS_THEN_DESCENDING
+}
+
+table SortKey {
+  value: Expression (required);
+  ordering: Ordering = ASCENDING_THEN_NULLS;
+}
+
+/// The contents of Relation.options will be OrderByOptions
+/// if Relation.operation = CanonicalOperation::OrderBy
+table OrderByOptions {
+  /// Define sort order for rows of output.
+  /// Keys with higher precedence are ordered ahead of other keys.
+  keys: [SortKey] (required);
+}
+
+/// The contents of Relation.options will be LimitOptions
+/// if Relation.operation = CanonicalOperation::Limit
+table LimitOptions {
+  /// Set the maximum number of rows of output.
+  count: long;
+}
+
+/// The contents of Relation.options will be CommonOptions
+/// if Relation.operation = CanonicalOperation::Common
+table CommonOptions {
+  /// Commons (CTEs in SQL) allow assigning a name to a stream
+  /// of data and reusing it, potentially multiple times and
+  /// potentially recursively.
+  name: string (required);
+}
+
+/// The contents of Relation.options will be UnionOptions
+/// if Relation.operation = CanonicalOperation::Union
+table UnionOptions {
+  /// For simplicity, all rows from any input to a union relation
+  /// will always be concatenated into a single output- establishing
+  /// uniqueness of output rows is deferred to other relations.
+}
+
+/// The contents of Relation.options will be LiteralOptions
+/// if Relation.operation = CanonicalOperation::Literal
+table LiteralOptions {
+  /// The columns of this literal relation.
+  columns: [Literal] (required);
+}
+
+/// A specification of a query.
+table Plan {
+  /// One or more output relations.
+  sinks: [Relation] (required);
```

Comment:
What is the interpretation of multiple outputs (versus a single output)?

Reply:
This is a difference in execution model for this proposal (for which `rpc_service Interactive` is provided semi-didactically to clarify).

If the root_type is a TableExpr which evaluates to batches, that implies that there is a channel open between the consumer and the producer along which those batches can be returned. This is frequently the case but in general I think we'll want to be able to express execution plans which don't rely on an interactive channel. For example: Plans generated as fragments of larger plans in distributed execution or plans which represent an ETL job.

Therefore I think it's preferable that a Plan explicitly include the destination for all batches, even if that will quite commonly be `operation=InteractiveOutput` (just pipe them back to the user).

Only a single instance of InteractiveOutput is permitted in a Plan (so no interactive producer will need to deinterleave batches piped back along the interactive channel, which I *think* was your concern here). However any number of other sinks are permitted. For example, a Plan may specify that one set of batches be streamed to `tcp://somehost.com:890` for consumption by a service on that host and an unfiltered superset of those batches cached locally into `file://tmp/cache/my_query` for debugging.